File: 0785-is-graph-bipartite/0785-is-graph-bipartite.js
function dfs(graph, node, colors, color) {
  colors[node] = color;
  for (const neighbor of graph[node]) {
    if (colors[neighbor] === color) {
      return false;
    }
    if (!colors[neighbor]) {
      if (!dfs(graph, neighbor, colors, -color)) {
        return false;
      }
    }
  }
  return true;
}

isBipartite = function(graph) {
  const n = graph.length;
  const colors = new Array(n).fill(0);  
  for (let i = 0; i < n; i++) {
    if (!colors[i]) {
      if (!dfs(graph, i, colors, 1)) {
        return false;
      }
    }
  }
  return true;
}

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n=len(graph)
        col=[-1]*n
        for i in range(n):
            if col[i]!=-1:
                continue

            q=deque()
            q.append((i,0))
            while q:
                node,color=q.popleft()
                if col[node]==-1:
                    col[node]=color
                    for nx in graph[node]:
                        q.append((nx,color^1))


                if col[node]!=color:
                    return False

        return True        